### PR TITLE
Leave single invocation of ar

### PR DESCRIPTION
**Pull request by sergv**
Fix for #4550.

Please include the following checklist in your PR:

* [X] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [X] Any changes that could be relevant to users have been recorded in the changelog.
* [X] The documentation has been updated, if necessary.
* [X] Old `DeterministicAr` test split in two to test new and old paths.

### CODE REVIEW

File: Cabal/Distribution/Simple/Compiler.hs [hunk 1]
```diff
@@ -339,6 +340,10 @@ libraryDynDirSupported comp = case compilerFlavor comp of
  where
   v = compilerVersion comp
 
+-- | Does this compiler's "ar" command supports @file-style arguments
+arAtFileArgumentSupported :: Compiler -> Bool
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Compiler.hs]: I suggest that we standardise on calling them "response files" in our codebase, since there's existing precedent (`haddockSupportsResponseFiles`).

File: Cabal/Distribution/Simple/Program/Ar.hs [hunk 2]
```diff
@@ -83,10 +86,29 @@ createArLibArchive verbosity lbi targetPath files = do
...
+        simpleArgs ++ extraArgs ++ ['@' : atFile]
+
+  if oldVersionManualOverride || atArgumentNotSupported
+    then
+      sequence_
         [ runProgramInvocation verbosity inv
         | inv <- multiStageProgramInvocation
                    simple (initial, middle, final) files ]
+    else
+      withTempFile tmpDir "object-files.txt" $ \argsPath argsHandle -> do
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Ar.hs]: Can we use `withTempFileEx` here, so that `--keep-tmp-files` would work?

**Comment by 23Skidoo** [comment]
There's also a minor merge conflict.

File: Cabal/Distribution/Simple/Program/Ar.hs (hunk 2 above)
Comment (sergv) [inline Cabal/Distribution/Simple/Program/Ar.hs]: I'm not sure there's a `--keep-tmp-files` flag in cabal-install. There's only `--keep-temp-files` in `cabal haddock`, but that's not relevant I reckon.

Surely I can put the `Ex` version there, but unfortunately I don't see what to pass to it except `defaultTempFileOptions`, which has temp files disabled unconditionally.

File: Cabal/Distribution/Simple/Compiler.hs (hunk 1 above)
Comment (sergv) [inline Cabal/Distribution/Simple/Compiler.hs]: Good point, I'll rename everything and amend the commits.

**Comment by sergv** [comment]
Actually, there's windows test failure and it's not random. My first version didn't account for response file format, which expects a bunch of whitespace-separated arguments. Thus, without quoting filenames with whitespace in them will not be processed correctly and Windows will be broken because it uses backslashes as path separators which are interpreted as quoting in the response file syntax. Similar issues were discussed in ghc when response files were added. The discussion in https://phabricator.haskell.org/D1158 is relevant, but I've summarized it already - the bottom line is that response files have each file name quoted and surrounded in quotes (double or single).

**Comment by sergv** [comment]
Updated, should be mergeable if tests pass.

File: Cabal/Distribution/Simple/Program/Ar.hs (hunk 2 above)
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Ar.hs]: OK. We should create a ticket about adding `--keep-tmp-files` to `GlobalFlags`.

File: Cabal/Distribution/Simple/Program/Ar.hs [hunk 3]
```diff
@@ -83,10 +86,29 @@ createArLibArchive verbosity lbi targetPath files = do
...
+        simpleArgs ++ extraArgs ++ ['@' : atFile]
+
+  if oldVersionManualOverride || responseArgumentsNotSupported
+    then
+      sequence_
         [ runProgramInvocation verbosity inv
         | inv <- multiStageProgramInvocation
                    simple (initial, middle, final) files ]
+    else
+      withTempFile tmpDir ".rsp" $ \responeFilePath responseFileHandle -> do
```
Comment (ezyang) [inline Cabal/Distribution/Simple/Program/Ar.hs]: s/responeFilePath/responseFilePath/

File: Cabal/Distribution/Simple/Program/Ar.hs [hunk 4]
```diff
@@ -97,9 +119,26 @@ createArLibArchive verbosity lbi targetPath files = do
...
+-- Namely, on Windows standard path separator is backspace but in @ar@
+-- response file format it's escape sequence. Therefore backslashes must
+-- be escaped as well.
+escapeFileName :: FilePath -> FilePath
+escapeFileName = concatMap escapeChar
+  where
+    escapeChar :: Char -> String
+    escapeChar '"'  = "\\\""
+    escapeChar '\\' = "\\\\"
+    escapeChar c    = [c]
```
Comment (ezyang) [inline Cabal/Distribution/Simple/Program/Ar.hs]: Comparing with https://git.haskell.org/ghc.git/commitdiff/296bc70b5ff6c853f2782e9ec5aa47a52110345e it seems you also need to handle newlines. Also, I am a bit perplexed: it doesn't look like you are ever wrapping this with quotes.

It would be really helpful to have a test case for these edge cases.

Comment (sergv) [inline Cabal/Distribution/Simple/Program/Ar.hs]: That's good catch, however it seems unplausible that file name will contain newlines. I'm not 100% sure that `ar` will like those newlines either.

File: Cabal/Distribution/Simple/Program/Ar.hs (hunk 3 above)
Comment (sergv) [inline Cabal/Distribution/Simple/Program/Ar.hs]: Good catch, thanks!

File: Cabal/Distribution/Simple/Program/Ar.hs (hunk 4 above)
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Ar.hs]: We have some code for producing response files in `D.S.Haddock` already, maybe we can consolidate it all under, say, `D.S.Program.ResponseFile`?

Comment (ezyang) [inline Cabal/Distribution/Simple/Program/Ar.hs]: I'd advise against it: the implementations of the response file parsers are different so it logically makes sense for there to be two implementations to (possibly) need to be different.

Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Ar.hs]: Maybe we can have both impls in a single file at least? I envision something like `withResponseFile (defaultResponseFileOptions { responseFileEscapingStyle = ArEscaping }) args $ \fname -> ...`

Comment (ezyang) [inline Cabal/Distribution/Simple/Program/Ar.hs]: Well, the code in any case is pretty short...

Comment (sergv) [inline Cabal/Distribution/Simple/Program/Ar.hs]: I think it would be best to unify the two under single implementation in `Distribution.Simple.Program.ResponseFile`. Haddock's module has following comment

```
      -- Support a gcc-like response file syntax.  Each separate
      -- argument and its possible parameter(s), will be separated in the
      -- response file by an actual newline; all other whitespace,
      -- single quotes, double quotes, and the character used for escaping
      -- (backslash) are escaped.  The called program will need to do a similar
      -- inverse operation to de-escape and re-constitute the argument list.
```

which implies that it tries to match gcc escaping rules and that's precisely what we're trying to achieve for `ar`, and possibly `ld`.

**Comment by sergv** [comment]
A couple of fixes, as discussed previously (can squash them later):
- I factored out response file handling into `Distribution.Simple.Program.ResponseFile`. The handling is taken from current haddock response file handling and should handle even newlines in file names
- I made `ld` invocation that creates library for ghci also use response files via heuristic mentioned in #4550 
- I renamed `--ar-does-not-support-response-files` argument to `configure` and `install` into `--disable-response-files` since it now affects both `ar` and `ld`. It does not affect haddock as it uses haddocks' version to detect whether response files are supported, which is pretty reliable

**Comment by sergv** [comment]
AppVeyor fails with unclear error message when running tests:
```
dist\build\cabal-tests\cabal-tests.exe -j3
threads: 3
tests to run: 186
PackageTests\AllowOlder\cabal.test.hs                                                                             SKIP (1.53s)
PackageTests\AllowNewer\cabal.test.hs                                                                             SKIP (1.73s)
Shutting down GHCi sessions (please be patient)...
Unexpected failure on GHCi exit: fd:3: hClose: resource vanished (Broken pipe)
cabal-tests.exe: fd:3: hClose: resource vanished (Broken pipe)
Command exited with code 1
```

Is it my changes or some intermittent failure?

**Comment by ezyang** [comment]
It shouldn't do that. I restarted the build.

**Comment by 23Skidoo** [comment]
> AppVeyor fails with unclear error message when running tests:
>
> [...]

This is #4618.

**Comment by sergv** [comment]
The build is green. Should we merge?

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Comment by sergv** [comment]
Great, thanks! :+1:
